File: project-euler/003/problem003.exs
#!/usr/bin/env elixir
defmodule Problem003 do
  def prime?(2), do: true
  def prime?(n) do
    upper_bound = n |> :math.sqrt
                    |> Float.ceil
                    |> trunc
    not_prime = 2..upper_bound
                |> Enum.any?(fn(x) -> rem(n, x) == 0 end)
    !not_prime
  end

  def factors(n) do
    1..n
    |> Enum.filter(fn(x) -> rem(n, x) == 0 end)
  end

  def solve do
    factors(600851475143)
    |> Enum.filter(&prime?/1)
    |> Enum.max
  end
end

IO.puts Problem003.solve

File: project-euler/003/problem003.rb
#!/usr/bin/env ruby
class Integer
  def prime?
    return false unless self > 1
    (2..Math.sqrt(self)).any? do |x|
      next if x == self
      return false if self % x == 0
    end
    true
  end

  def factors
    (1..Math.sqrt(self)).select { |x| self % x == 0 }.reduce([]) do |list, number|
      list << number
      list << self / number unless self == number
      list
    end
  end
end

puts 600851475143.factors.select { |x| x.prime? }.max